cmd/kubeadm/app/cmd/upgrade/node.go

```diff
@@ -19,221 +19,253 @@ package upgrade
+		Use:   "node",
+		Short: "Upgrade commands for a node in the cluster",
 		Run: func(cmd *cobra.Command, args []string) {
-			err := RunUpgradeNodeConfig(flags)
+			c, err := nodeRunner.InitData(args)
 			kubeadmutil.CheckErr(err)
-		},
-	}
 
-	options.AddKubeConfigFlag(cmd.Flags(), &flags.kubeConfigPath)
-	cmd.Flags().BoolVar(&flags.dryRun, options.DryRun, flags.dryRun, "Do not change any state, just output the actions that would be performed.")
-	cmd.Flags().StringVar(&flags.kubeletVersionStr, "kubelet-version", flags.kubeletVersionStr, "The *desired* version for the kubelet after the upgrade.")
-	return cmd
-}
-
-// NewCmdUpgradeControlPlane returns the cobra.Command for upgrading the controlplane instance on this node
-func NewCmdUpgradeControlPlane() *cobra.Command {
+			data := c.(*nodeData)
+			switch data.IsControlPlaneNode() {
+			case true:
+				fmt.Println("[upgrade] Control-plane node detected")
+			case false:
+				fmt.Println("[upgrade] Worker node detected")
+			}
 
-	flags := &controlplaneUpgradeFlags{
-		kubeConfigPath:   constants.GetKubeletKubeConfigPath(),
-		advertiseAddress: "",
-		etcdUpgrade:      true,
-		renewCerts:       true,
-		dryRun:           false,
+			err = nodeRunner.Run(args)
+			kubeadmutil.CheckErr(err)
+		},
+		Args: cobra.NoArgs,
 	}
 
-	cmd := &cobra.Command{
-		Use:     "experimental-control-plane",
-		Short:   "Upgrade the control plane instance deployed on this node. IMPORTANT. This command should be executed after executing `kubeadm upgrade apply` on another control plane instance",
-		Long:    upgradeNodeConfigLongDesc,
-		Example: upgradeNodeConfigExample,
-		Run: func(cmd *cobra.Command, args []string) {
+	// adds flags to the node command
+	// flags could be eventually inherited by the sub-commands automatically generated for phases
+	addUpgradeNodeFlags(cmd.Flags(), nodeOptions)
 
-			if flags.nodeName == "" {
-				klog.V(1).Infoln("[upgrade] found NodeName empty; considered OS hostname as NodeName")
-			}
-			nodeName, err := node.GetHostname(flags.nodeName)
-			if err != nil {
-				kubeadmutil.CheckErr(err)
-			}
-			flags.nodeName = nodeName
+	// initialize the workflow runner with the list of phases
+	nodeRunner.AppendPhase(phases.NewControlPlane())
+	nodeRunner.AppendPhase(phases.NewKubeletConfigPhase())
 
-			if flags.advertiseAddress == "" {
-				ip, err := configutil.ChooseAPIServerBindAddress(nil)
-				if err != nil {
-					kubeadmutil.CheckErr(err)
-					return
-				}
+	// sets the data builder function, that will be used by the runner
+	// both when running the entire workflow or single phases
+	nodeRunner.SetDataInitializer(func(cmd *cobra.Command, args []string) (workflow.RunData, error) {
+		return newNodeData(cmd, args, nodeOptions)
+	})
 
-				flags.advertiseAddress = ip.String()
-			}
+	// binds the Runner to kubeadm upgrade node command by altering
+	// command help, adding --skip-phases flag and by adding phases subcommands
+	nodeRunner.BindToCommand(cmd)
```

Comment:
in an ideal world `kubeadm upgrade apply` should have been the UX that works all use cases:
- primary CP
- secondary CP
- worker
(as long as we can detect primary vs secondary)

Reply:
@neolit123 that's a nice UX suggestion, but a bit hard and unreliable to implement in my opinion. The `primary CP` does not need to be invoked on the `kubeadm init` node. It just needs to be done on a CP node only once. Then it should store somewhere a result of this operation and allow for the same command to detect it on other nodes.
This may be unreliable, hence we should always have the manual interface as a safe backup.

Reply:
I'm not sure it is a good idea, but this is something out of scope of this PR

Reply:
clearly out of scope of this PR. just saying that the `apply` vs `node` is not exactly intuitive.

Reply:
@rosti, @neolit123, checking the logic for `kubeadm upgrade apply` I'm wondering what is preventing it from being idempotent, assuming the same version is passed to `kubeadm upgrade apply` on all control plane nodes.

* `PerformControlPlaneUpgrade` -> `PerformStaticPodUpgrade` is trivial and always happens.
* `PerformPostUpgradeTasks` from what I could see should be safe to re-run.

Of course is not a "supported" way of upgrading, but I fail to see what is preventing us from doing so. In general, once that we have run `kubeadm upgrade apply` once in one control plane, we'll have upgraded the `ClusterConfiguration` with the new Kubernetes version, and in secondary control planes we could know from what version we are trying to update to what version (present in the `kubeadm-config` configmap in the cluster configuration).

Unsure if I'm missing something...

Reply:
> once that we have run kubeadm upgrade apply once in one control plane, we'll have upgraded the ClusterConfiguration with the new Kubernetes version, and in secondary control planes we could know from what version we are trying to update to what version (present in the kubeadm-config configmap in the cluster configuration).

this is similar to what i was thinking without doing much research on this topic.